### DISCUSSION TITLE
Please help me with extention for SpanProcessor

### DISCUSSION THREAD

**Discussion by YonatanHunger** [open-telemetry/opentelemetry-java-instrumentation#5512] [Q&A] [answered]
I am trying to extend OpenTelemetry java agent and I don't see any indication it is trying to load my jar.
I am running the following cmd: 
```
java -javaagent:../src/main/resources/opentelemetry-javaagent.jar -Dotel.javaagent.configuration-file=../src/main/resources/agent-prp.properties -jar simple-service-1.0-SNAPSHOT-jar-with-dependencies.jar
```
my config file is (the attributes are working):
```
otel.javaagent.extensions=/Users/foo/source/simple-service/src/main/resources/span-processor-1.0-SNAPSHOT.jar
otel.resource.attributes=service.name=foooBarr
```
my extension is:
```
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import io.opentelemetry.context.Context;
import io.opentelemetry.sdk.common.CompletableResultCode;
import io.opentelemetry.sdk.trace.ReadWriteSpan;
import io.opentelemetry.sdk.trace.ReadableSpan;
import io.opentelemetry.sdk.trace.SpanProcessor;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
...

public class FooSpanProcessor implements SpanProcessor {
    private static final ObjectMapper objMapper = new ObjectMapper();
    private static final Logger log = LoggerFactory.getLogger(FooSpanProcessor.class);


    @Override
    public void onStart(Context parentContext, ReadWriteSpan span) {
        span.setAttribute("fooToken", FooProperties.INSTANCE.fooToken);
        span.setAttribute("service.name", FooProperties.INSTANCE.serviceName);
        span.setAttribute("runtime", FooProperties.INSTANCE.javaVersion);
        span.setAttribute("tracerVersion", "0.0.1");
        span.setAttribute("framework", FooProperties.INSTANCE.frameWork);
        span.setAttribute("envs", FooProperties.INSTANCE.environment);
        span.setAttribute("metaData", FooProperties.INSTANCE.metadata);
    }

    @Override
    public boolean isStartRequired() {
        return true;
    }

    @Override
    public void onEnd(ReadableSpan span) {
    }
    ....
```

I have added the configuration like this:
```
@AutoService(SdkTracerProviderConfigurer.class)
public class FooSdkTracerProviderConfigurer implements SdkTracerProviderConfigurer {
    @Override
    public void configure(SdkTracerProviderBuilder tracerProvider, ConfigProperties config) {
        tracerProvider.addSpanProcessor(new FooSpanProcessor());
    }
```
    
I don't see any indication that my extension is loaded, I don't see any of my parameters on the produced spans. can anybody help me?

**Comment by anuraaga** [comment, marked as answer]
Hi @YonatanHunger - due to an initialization ordering constraint, the `otel.javaagent.extensions` does not support being read from the properties file. Can you pass `-Dotel.javaagent.extensions=` on the command line and see if it helps?

**Reply by YonatanHunger** [reply]
my god, two days of my life. thank you so much!

**Comment by Yahire3388** [comment]
Hi @YonatanHunger @anuraaga ,

I am trying the same thing, i am using tomcat server and had passed java agent and its extension in bellow format : 
![16474275763641693017025707975818](https://user-images.githubusercontent.com/98525040/158573107-479b513e-a175-4814-9d6f-75610ef06540.jpg)

But still the traces are not visible.
@YonatanHunger , can you please share your jar what you had created so that i can confirm if the issue is with my jar or the setenv file of tomcat.

Thankyou

**Reply by mateuszrzeszutek** [reply]
Hey @Yahire3388 ,
Your screenshot is slightly hard to read, but are you perhaps using the `-` instead of `=` in a few places? Like `-Dotel.javaagent.extensions-D:/...` instead of `-Dotel.javaagent.extensions=D:/...`
If so, that's not a valid property syntax -- try using the `=` char between the property name and value instead.

**Reply by Yahire3388** [reply]
Hi @mateuszrzeszutek ,

Pfb, the setenv : 

set "CATALINA_OPTS=%CATALINA_OPTS%" -javaagent:D:/agent.jar -Dotel.javaagent.extensions=D:/demoextension.jar

It seems to be correct syntax.

Thanks

**Reply by Yahire3388** [reply]
Does it work with auto instrumentation? I am using auto instrumentation for tracing and just adding this extension in setenv thats it.

**Reply by mateuszrzeszutek** [reply]
Yes, the extension feature only works with the agent.
Can you perhaps paste the contents of the `setenv.bat` file here? As I mentioned before, it's pretty hard to read them from your picture, and maybe we could correct any mistakes that are in this file.

**Reply by Yahire3388** [reply]
Following is my setenv.bat :

set "CATALINA_OPTS=%CATALINA_OPTS% -javaagent:D:/Opentelemetry/opentelemetry-javaagent.jar -Dotel.javaagent.extensions=D:/apache-tomcat-9.0.14/demoextension.jar -Dotel.traces.exporter=zipkin -Djavax.net.debug=all -Dotel.metrics.exporter=none -Dotel.instrumentation.common.db-statement-sanitizer.enabled=true -Dotel.resource.attributes=service.name=OPENTELEMETRY,service.namespace=java -Dotel.exporter.zipkin.endpoint="http://localhost:9411/api/v2/spans""

**Reply by Yahire3388** [reply]
Hi @mateuszrzeszutek , @YonatanHunger , @anuraaga ,
following is my tomcat server logs which shows command line arguments it received : 
16-Mar-2022 18:08:21.945 INFO [main] org.apache.catalina.startup.VersionLoggerListener.log Command line argument: -javaagent:D:/Opentelemetry/opentelemetry-javaagent.jar
16-Mar-2022 18:08:21.946 INFO [main] org.apache.catalina.startup.VersionLoggerListener.log Command line argument: -Dotel.javaagent.experimental.extensions=D:/apache-tomcat-9.0.14/demoextension.jar

**Reply by Yahire3388** [reply]
@trask ,

FYI, please refer the discussion.
Please help me to get this issue resolve.

Thanks

**Reply by trask** [reply]
hi @Yahire3388! can you create a minimal repro of your issue in a public github repository, and then open a new issue here with instructions how to reproduce the issue you are seeing?

**Reply by Yahire3388** [reply]
Hi @trask 
 please refer https://github.com/Yahire3388/extensionToJavaAgent/issues/1 . I had created a repository with the extension code i am using and the steps i am performing to add extension.

Thanks

**Reply by trask** [reply]
thx @Yahire3388, can you create a new issue or new discussion also? it is harder for us to keep track when you comment across multiple old discussion threads, thx

**Reply by Yahire3388** [reply]
Hi @trask 

Please refer https://github.com/open-telemetry/opentelemetry-java-instrumentation/discussions/5647.
This is the new issue i had created in the discussion here. 

Thankyou
